java/drivers/driver-hazelcast4plus/src/main/java/com/hazelcast/simulator/tests/map/prunability/ScanByPrunedCompositePartialKeyBenchmark.java

```diff
@@ -0,0 +1,118 @@
+import com.hazelcast.map.IMap;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.DataSerializable;
+import com.hazelcast.simulator.hz.HazelcastTest;
+import com.hazelcast.simulator.hz.IdentifiedDataSerializablePojo;
+import com.hazelcast.simulator.test.annotations.Prepare;
+import com.hazelcast.simulator.test.annotations.Setup;
+import com.hazelcast.simulator.test.annotations.Teardown;
+import com.hazelcast.simulator.test.annotations.TimeStep;
+import com.hazelcast.simulator.worker.loadsupport.Streamer;
+import com.hazelcast.simulator.worker.loadsupport.StreamerFactory;
+import com.hazelcast.sql.SqlResult;
+import com.hazelcast.sql.SqlRow;
+import com.hazelcast.sql.SqlService;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Random;
+
+import static java.util.Arrays.asList;
+
+
+public class ScanByPrunedCompositePartialKeyBenchmark extends HazelcastTest {
+
+    // properties
+    // the number of map entries
+    public int entryCount = 10_000_000;
+
+    private IMap<KeyPojo, String> map;
+
+    @Setup
+    public void setUp() {
+        this.map = targetInstance.getMap(name);
+        final List<PartitioningAttributeConfig> attributeConfigs = asList(
+                new PartitioningAttributeConfig("a"),
+                new PartitioningAttributeConfig("b")
+        );
+        final MapConfig mapConfig = new MapConfig(name).setPartitioningAttributeConfigs(attributeConfigs);
+        targetInstance.getConfig().addMapConfig(mapConfig);
+    }
+
+    @Prepare(global = true)
+    public void prepare() {
+        Streamer<KeyPojo, String> streamer = StreamerFactory.getInstance(map);
+
+        for (int i = 0; i < entryCount; i++) {
+            Integer iKey = i;
+            Long lKey = (long) i;
+            String v = String.format("%010d", iKey);
+            KeyPojo keyPojo = new KeyPojo(iKey, v, lKey);
+            streamer.pushEntry(keyPojo, v);
+        }
+        streamer.await();
+
+        SqlService sqlService = targetInstance.getSql();
+        String query = "CREATE EXTERNAL MAPPING IF NOT EXISTS " + name + " "
+                + "EXTERNAL NAME " + name + " "
+                + "        TYPE IMap\n"
+                + "        OPTIONS (\n"
+                + "                'keyFormat' = 'java',\n"
+                + "                'keyJavaClass' = 'com.hazelcast.simulator.tests.map.prunability.KeyPojo',\n"
+                + "                'valueFormat' = 'java',\n"
+                + "                'valueJavaClass' = 'java.lang.String'\n"
+                + "        )";
+
+        sqlService.execute(query);
+    }
+
+    @TimeStep
+    public void timeStep() throws Exception {
+        SqlService sqlService = targetInstance.getSql();
+        String query = "SELECT __key, this FROM " + name + " WHERE \"__key.a\"= ? AND \"__key.b\"= ?";
+        String keyMatch = String.format("%010d", new Random().nextInt(entryCount));
```

Comment:
you will get messy results because some queries will execute only on coordinator and some on coordinator+additional member, depending on the partition assignment. It will be harder to analyse such results, ofc they will give some average, but p99 will be meaningless.

Reply:
>you will get messy results because some queries will execute only on coordinator and some on coordinator+additional member

This is especially impactful for constant key tests.

Reply:
The main aim behind this benchmark is to answer a question - "are queries with partition pruning applied faster than **the same queries** on **the same setup**?". I assume, that given setup answers this question.

Reply:
in order to answer such question it would be sufficient to run some queries in a loop locally. Our other goal is to publish those results and for that we should have a reasonable methodology.

Reply:
as discussed, it is impossible to control this now, because the query is executed on random member unless partition aware client is used, but it currently does not integrate with partition pruning